bsp/Src/CAN.c

```diff
@@ -0,0 +1,801 @@
+    init.Alternate = GPIO_AF9_CAN2;
+    HAL_GPIO_Init(GPIOB, &init);
+
+    init.Pin = GPIO_PIN_13;
+    init.Mode = GPIO_MODE_AF_PP;
+    init.Pull = GPIO_NOPULL;
+    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    init.Alternate = GPIO_AF9_CAN2;
+    HAL_GPIO_Init(GPIOB, &init);
+
+    // enable interrupts
+    HAL_NVIC_SetPriority(CAN2_TX_IRQn, 5, 0);
+    HAL_NVIC_EnableIRQ(CAN2_TX_IRQn);
+    HAL_NVIC_SetPriority(CAN2_RX0_IRQn, 5, 0);
+    HAL_NVIC_EnableIRQ(CAN2_RX0_IRQn);
+  }
+  #endif /* CAN2 */
+
+  // CAN3
+  #ifdef CAN3
+  else if (hcan->Instance == CAN3) {
+    // enable clocks
+    __HAL_RCC_CAN3_CLK_ENABLE();
+    __HAL_RCC_GPIOA_CLK_ENABLE();
+
+    /* enable gpio
+    PA8  -> CAN3_RX
+    PA15 -> CAN3_TX
+    */
+    init.Pin = GPIO_PIN_8;
+    init.Mode = GPIO_MODE_AF_PP;
+    init.Pull = GPIO_PULLUP;
+    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    init.Alternate = GPIO_AF11_CAN3;
+    HAL_GPIO_Init(GPIOA, &init);
+
+    init.Pin = GPIO_PIN_15;
+    init.Mode = GPIO_MODE_AF_PP;
+    init.Pull = GPIO_NOPULL;
+    init.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    init.Alternate = GPIO_AF11_CAN3;
+    HAL_GPIO_Init(GPIOA, &init);
+
+    // enable interrupts
+    HAL_NVIC_SetPriority(CAN3_TX_IRQn, 5, 0);
+    HAL_NVIC_EnableIRQ(CAN3_TX_IRQn);
+    HAL_NVIC_SetPriority(CAN3_RX0_IRQn, 5, 0);
+    HAL_NVIC_EnableIRQ(CAN3_RX0_IRQn);
+  }
+  #endif /* CAN3 */
+}
+
+// CAN MSP deinit
+void HAL_CAN_MspDeInit(CAN_HandleTypeDef* hcan) {
+  // CAN1
+  if (hcan->Instance == CAN1) {
+    // disable clocks
+    __HAL_RCC_CAN1_CLK_DISABLE();
+
+    /* disable gpio
+    PA11 -> CAN1_RX
+    PA12 -> CAN1_TX
+    */
+    HAL_GPIO_DeInit(GPIOA, GPIO_PIN_11);
+    HAL_GPIO_DeInit(GPIOA, GPIO_PIN_12);
+
+    // disable interrupts
+    HAL_NVIC_DisableIRQ(CAN1_TX_IRQn);
+    HAL_NVIC_DisableIRQ(CAN1_RX0_IRQn);
+  }
+
+  // CAN2
+  #ifdef CAN2
+  else if (hcan->Instance == CAN2) {
+    // disable clocks
+    __HAL_RCC_CAN2_CLK_DISABLE();
+
+    /* disable gpio
+    PB12 -> CAN2_RX
+    PB13 -> CAN2_TX
+    */
+    HAL_GPIO_DeInit(GPIOB, GPIO_PIN_12);
+    HAL_GPIO_DeInit(GPIOB, GPIO_PIN_13);
+
+    // disable interrupts
+    HAL_NVIC_DisableIRQ(CAN2_TX_IRQn);
+    HAL_NVIC_DisableIRQ(CAN2_RX0_IRQn);
+  }
+  #endif /* CAN2 */
+
+  // CAN3
+  #ifdef CAN3
+  else if (hcan->Instance == CAN3) {
+    // disable clocks
+    __HAL_RCC_CAN3_CLK_DISABLE();
+
+    /* disable gpio
+    PA8  -> CAN3_RX
+    PB15 -> CAN3_TX
+    */
+    HAL_GPIO_DeInit(GPIOA, GPIO_PIN_8);
+    HAL_GPIO_DeInit(GPIOA, GPIO_PIN_15);
+
+    // disable interrupts
+    HAL_NVIC_DisableIRQ(CAN3_TX_IRQn);
+    HAL_NVIC_DisableIRQ(CAN3_RX0_IRQn);
+  }
+  #endif /* CAN3 */
+}
+
+can_status_t can_init(CAN_HandleTypeDef* handle, CAN_FilterTypeDef* filter) {
+  // CAN1
+  if (handle->Instance == CAN1) {
+    // init semaphores
+    can1_send_semaphore =
+        xSemaphoreCreateBinaryStatic(&can1_send_semaphore_buffer);
+    xSemaphoreGive(can1_send_semaphore);
```

Comment:
any error handling for these semaphore calls?

Reply:
They give no error I believe. All they do is populate the empty data structures

Reply:
Wait what about xSemaphoreGive?

Reply:
We can assume it is fine since I just created the semaphore. But also, I'm taking out the semaphores, so it doesn't matter anymore